src/passes/pass.cpp

```diff
@@ -138,19 +138,23 @@ void PassRunner::addDefaultGlobalOptimizationPasses() {
 }
 
 void PassRunner::run() {
-  if (options.debug) {
+  // BINARYEN_PASS_DEBUG is a convenient commandline way to log out the toplevel passes, their times,
+  //                     and validate between each pass.
+  //                     (we don't recurse pass debug into sub-passes, as it doesn't help anyhow and
+  //                     also is bad for e.g. printing which is a pass)
+  static const int passDebug = getenv("BINARYEN_PASS_DEBUG") ? atoi(getenv("BINARYEN_PASS_DEBUG")) : 0;
+  if (!isNested() && (options.debug || passDebug)) {
```

Comment:
Splitting it up is separate if we keep `isNested` I think.

I think we want `isNested` iff we don't have a `NestedPassRunner` subclass. Most directly: if we use `isNested` for dispatching to nested or not-nested code, the code for the nested case would live in `PassRunner`, and not `NestedPassRunner`. In the limit, the only code in `NestedPassRunner` is an override of `isNested`, which smells like we're using the wrong tool. My argument is basically, if the purpose behind the subclass is to put nesting-specific logic there, and deciding which passes to run is nesting-specific, then that's where that logic should live.
Subclassing can be a very heavy hammer, as it constrains and enforces design in various ways. For example if we introduce another kind of `PassRunner`, say a `FooPassRunner`, can a `FooPassRunner` be nested?
I guess I'm asking, why promote the concept of nesting into the class hierarchy in the first place?